src/librustc/hir/map/mod.rs

```diff
@@ -633,6 +708,17 @@ impl<'hir> Map<'hir> {
         self.find_entry(id).and_then(|x| x.parent_node()).unwrap_or(id)
     }
 
+    // FIXME(@ljedrz): replace the NodeId variant
+    pub fn get_parent_node_by_hir_id(&self, id: HirId) -> HirId {
+        if self.dep_graph.is_fully_enabled() {
+            let hir_id_owner = id.owner;
+            let def_path_hash = self.definitions.def_path_hash(hir_id_owner);
+            self.dep_graph.read(def_path_hash.to_dep_node(DepKind::HirBody));
+        }
+
+        self.find_entry_by_hir_id(id).and_then(|x| x.parent_hir()).unwrap_or(id)
+    }
+
```

Comment:
This should just convert the `HirId` to a `NodeId` and call the regular `get_parent_node` method.